pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -234,9 +234,35 @@ func (i *Instances) List(name_filter string) ([]string, error) {
 	return ret, nil
 }
 
-func getServerByName(client *gophercloud.ServiceClient, name string) (*servers.Server, error) {
+func getServerByHostName(client *gophercloud.ServiceClient, hostname string) (*servers.Server, error) {
+	// Depending on OpenStack setup, 'hostname' may include also domain name, while
+	// the server ID does not have it.
+	// Try to find server with ID which matches "^shortname(.domainname)?$"
+
+	domain := ""
+	shortname := hostname
+	if net.ParseIP(hostname) == nil {
+		// hostname is not an IP address
+		components := strings.SplitN(hostname, ".", 2)
+		shortname = components[0]
+		if len(components) > 1 {
+			domain = components[1]
+		}
+	}
+
+	var nameRegExp string
+	if domain == "" {
+		nameRegExp = fmt.Sprintf("^%s$", regexp.QuoteMeta(hostname))
+	} else {
+		nameRegExp = fmt.Sprintf("^%s(\\.%s)?$", regexp.QuoteMeta(shortname), regexp.QuoteMeta(domain))
```

Comment:
Could a server exist with just the shortname AND one exist with the full name? What would happen in that scenario?